daemon/build.go

```diff
@@ -104,13 +105,19 @@ func (daemon *Daemon) pullForBuilder(ctx context.Context, name string, authConfi
 // Every call to GetImageAndReleasableLayer MUST call releasableLayer.Release() to prevent
 // leaking of layers.
 func (daemon *Daemon) GetImageAndReleasableLayer(ctx context.Context, refOrID string, opts backend.GetImageAndLayerOptions) (builder.Image, builder.ReleaseableLayer, error) {
-	if !opts.ForcePull {
-		image, _ := daemon.GetImage(refOrID)
+	id, _ := daemon.GetImageID(refOrID)
+	refIsID := id.String() == refOrID // detect if ref is an ID to skip pulling
```

Comment:
The logic in `GetImageID` just tries to find a local image match based on checks with different priorities. It doesn't do know anything about if that ref may exist in registry or not. "ref or ID" is determined by 1) unambiguous full-length match 2) searching prefix from local imagestore.

It just depends what we consider `--pull` semantics to be. If it is `docker pull && docker run` then current is correct. If it is `docker pull || true && docker run` then it has some potential attack vectors if the registry is unreachable.